benches/dapol.rs

```diff
@@ -1,175 +1,438 @@
-use core::fmt::Debug;
-use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion, SamplingMode};
+mod setup;
+
+use criterion::{criterion_group, criterion_main, SamplingMode};
+use criterion::{BenchmarkId, Criterion};
+use dapol::accumulators::{NdmSmt, NdmSmtSecrets};
+use iai_callgrind::{black_box, library_benchmark, library_benchmark_group, main};
+
+use std::path::PathBuf;
+use std::str::FromStr;
+
 use dapol::{
-    utils::get_secret, Dapol, DapolNode, RangeProofPadding, RangeProofSplitting, RangeProvable,
-    RangeVerifiable,
+    Entity, EntityId, Height, MaxThreadCount, Secret,
 };
-use digest::Digest;
-use rand::{distributions::Uniform, thread_rng, Rng};
-use smtree::{
-    index::TreeIndex,
-    traits::{ProofExtractable, Rand, Serializable, TypeName},
-};
-use std::time::Duration;
 
-// CONSTANTS
-// ================================================================================================
+use setup::{NUM_USERS, TREE_HEIGHTS};
 
-const TREE_HEIGHTS: [usize; 3] = [16, 24, 32];
-const NUM_LEAVES: [usize; 3] = [1024, 2048, 4096];
-
-// BENCHMARKS
+// BENCHMARKS: CRITERION
 // ================================================================================================
 
-fn build_dapol(c: &mut Criterion) {
+fn bench_build_tree(c: &mut Criterion) {
     let mut group = c.benchmark_group("build");
     group.sample_size(10);
     group.sampling_mode(SamplingMode::Flat);
-    group.measurement_time(Duration::from_secs(20));
-
-    // bench tree height = 16
-    let tree_height = 16;
-    for &num_leaves in NUM_LEAVES.iter() {
-        let items = build_item_list(num_leaves, tree_height);
-        group.bench_function(BenchmarkId::new("height_16", num_leaves), |bench| {
-            bench.iter(|| {
-                // we bench range proof padding only because building a tree does not depend on
-                // the type of range proof we do
-                build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height)
-            });
-        });
-    }
 
-    // bench tree height = 32
-    let tree_height = 32;
-    for &num_leaves in NUM_LEAVES.iter() {
-        let items = build_item_list(num_leaves, tree_height);
-        group.bench_function(BenchmarkId::new("height_32", num_leaves), |bench| {
-            bench.iter(|| {
-                // we bench range proof padding only because building a tree does not depend on
-                // the type of range proof we do
-                build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height)
-            });
-        });
-    }
+    dapol::initialize_machine_parallelism();
 
-    group.finish();
-}
+    let max_thread_count = MaxThreadCount::default().get_value();
 
-fn generate_proof(c: &mut Criterion) {
-    let mut group = c.benchmark_group("prove");
-    group.sample_size(10);
+    let mut thread_counts: Vec<u8> = Vec::new();
+
+    if max_thread_count <= 8 {
```

Comment:
A much cleaner way to do this would be:
```rust
    let k = u64::BITS - max_thread_count.leading_zeros() - 1; // floor(log_2(max_thread_count))
    let lower_bound = 1u64 << k-1; // 2^k
    let upper_bound = 1u64 << k; // 2^(k+1)
    let step = 1usize << k-3; // 2^(k-2)
    for i in (lower_bound..upper_bound).step_by(step) {
        thread_counts.push(i);
    }
    thread_counts.push(max_thread_count);
```

It's doesn't give exactly the same numbers but that's fine.